### PR TITLE
Feature/25 setup socketio

### PR DESCRIPTION
**Pull request by wayangalihpratama**
- #25 

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207662929082466
  - https://app.asana.com/0/0/1207380167122126

### CODE REVIEW

**Review by dedenbangkit — CHANGES_REQUESTED** [review]
The faker seeder seems invalid on seeding fake phone number @wayangalihpratama

**Comment by wayangalihpratama** [comment]
> The faker seeder seems invalid on seeding fake phone number @wayangalihpratama

- I solved the socketio testing problem by using AsyncMock on test case.

File: packages/Akvo_rabbitmq_client/src/Akvo_rabbitmq_client/queue_message_util.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        if media is None:
+            media = []
+        if context is None:
+            context = []
+        if transformation_log is None:
+            transformation_log = [body]
+
+        message = {
+            "conversation_envelope": {
+                "message_id": str(uuid4()),
```
Comment (kjkoster) [inline packages/Akvo_rabbitmq_client/src/Akvo_rabbitmq_client/queue_message_util.py:25]: The message ID should not be generated server-side, as stated in https://docs.google.com/document/d/12koOL2T5Wstga0EBEwicDZgKq_Y9-dVpOP0OKin5xL8/edit#heading=h.u16r6w9czmp3 I think this is important for deduplication.

File: packages/Akvo_rabbitmq_client/tests/test_queue_message_util.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import unittest
+from uuid import UUID
+from Akvo_rabbitmq_client import queue_message_util
+
+
+class TestQueueMessageUtil(unittest.TestCase):
```
Comment (kjkoster) [inline packages/Akvo_rabbitmq_client/tests/test_queue_message_util.py:6]: Good to see tests!

**Review by kjkoster — CHANGES_REQUESTED** [review]

File: backend/core/socketio_config.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        body=json.dumps(msg),
+        routing_key=RABBITMQ_QUEUE_USER_CHATS)
+
+
+async def chat_replies_callback(body: str):
+    logger.info(f"Server received chat_replies_callback: {body}")
+    await sio_server.emit(
+        'chats',
+        {
+            "phone": "+628123456789",
```
[the remark this answers is not in the document]
Comment (wayangalihpratama) [reply backend/core/socketio_config.py:53]: Hi @kjkoster , 
This phone number will be fetch from user/client, so in my opinion we doesn't need to  move this into `.env`

File: backend/core/socketio_config.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """Track user disconnection"""
+    logger.info('User disconnected')
+
+
+@sio_server.on('chats')
+async def chat_message(sid, msg):
+    """Receive a chat message and send to all clients"""
+    RABBITMQ_QUEUE_USER_CHATS = os.getenv('RABBITMQ_QUEUE_USER_CHATS')
+    logger.info(f"Server received: {msg} {RABBITMQ_QUEUE_USER_CHATS}")
+    await sio_server.emit('chats', msg)
```
[the remark this answers is not in the document]
Comment (wayangalihpratama) [reply backend/core/socketio_config.py:42]: For this, Socket.IO will send the message back to the frontend. I am still working on it, so we don't need to send the message back from the frontend socket. I tried removing the code on line 42, but the chats did not appear in the frontend.

File: frontend/src/components/chats/ChatWindow.js [hunk 5]
```diff
@@ -46,12 +68,56 @@ const ChatWindow = () => {
...
     if (message.trim()) {
+      socket.timeout(5000).emit("chats", { phone: "+628123456789", message });
       // Implement your send message logic here
       setMessage(""); // Clear the textarea after sending
       textareaRef.current.style.height = "auto"; // Reset the height after sending
     }
   };
 
+  const renderChatMessages = () => {
+    return chats.map((c, ci) => {
```
[the remark this answers is not in the document]
Comment (wayangalihpratama) [reply frontend/src/components/chats/ChatWindow.js:79]: for now we are not using the ordering yet.

File: packages/Akvo_rabbitmq_client/src/Akvo_rabbitmq_client/queue_message_util.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+                "conversation_id": str(uuid4()),
+                "client_id": client_id,
+                "user_id": user_id
```
[the remark this answers is not in the document]
Comment (wayangalihpratama) [reply packages/Akvo_rabbitmq_client/src/Akvo_rabbitmq_client/queue_message_util.py:28]: Sorry, I don't get you, 
I'm thinking to change the client_id and user_id into phone_number

File: backend/core/socketio_config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wayangalihpratama) [reply backend/core/socketio_config.py:42]: - we can use `indexedDB` in FE to store the chats locally.
- generate UUID in the FE if the message started from user/extension officer, else we will use MessageSid from twilio/slack.

File: frontend/src/components/chats/ChatWindow.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (wayangalihpratama) [reply frontend/src/components/chats/ChatWindow.js:79]: - use UTC time

[the remark this answers is not in the document]
Comment (kjkoster) [reply frontend/src/components/chats/ChatWindow.js:80]: See my updated description of the frontend component: https://docs.google.com/document/d/12koOL2T5Wstga0EBEwicDZgKq_Y9-dVpOP0OKin5xL8/edit#heading=h.ecgg5gdax48i

**Comment by wayangalihpratama** [comment]
Hi @kjkoster , 
I have tried to implement the queue message format on both the frontend and backend. However, some values are still static and will be updated later once we connect all the elements (Slack, Twilio, etc.).

**Review by kjkoster — APPROVED** [review]

**Pull request merged**